src/node_i18n.cc

```diff
@@ -185,13 +219,379 @@ static void ToASCII(const FunctionCallbackInfo<Value>& args) {
+  OPEN_CONVERTER(from_conv, *from_name, status);
+
+  ucnv_setSubstChars(to_conv, "?", 1, &status);
+  status = U_ZERO_ERROR;
+
+  limit = ts_obj_length * ucnv_getMaxCharSize(to_conv);
+  buf.AllocateSufficientStorage(ts_obj_length * ucnv_getMaxCharSize(to_conv));
+  target = *buf;
+
+  ucnv_convertEx(to_conv, from_conv,
+                 &target, target + limit,
+                 &data, ts_obj_data + ts_obj_length,
+                 NULL, NULL, NULL, NULL,
+                 true, true,
+                 &status);
+
+  if (U_SUCCESS(status)) {
+    len = target - *buf;
+    args.GetReturnValue().Set(
+      Buffer::Copy(env->isolate(), *buf, len).ToLocalChecked());
+    goto cleanup;
+  }
+
+ error:
+  env->ThrowError(u_errorName(status));
+
+ cleanup:
+  if (to_conv != nullptr)
+    ucnv_close(to_conv);
+  if (from_conv != nullptr)
+    ucnv_close(from_conv);
+}
+
+// Converts to UCS2 from ISO-8859-1 and US-ASCII
+// args[0] is the encoding to convert from
+// args[1] must be a buffer instance
+void ConvertToUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[1]);
+
+  Utf8Value name(env->isolate(), args[0]);
+
+  SPREAD_ARG(args[1], ts_obj);
+
+  UConverter* to_conv = nullptr;
+  UErrorCode status = U_ZERO_ERROR;
+  size_t len;
+
+  MaybeStackBuffer<UChar> buf;
+  buf.AllocateSufficientStorage(ts_obj_length);
+
+  OPEN_CONVERTER(to_conv, *name, status);
+
+  len = ucnv_toUChars(to_conv,
+                      *buf, ts_obj_length << 1,
+                      ts_obj_data, ts_obj_length,
+                      &status);
+
+  if (U_SUCCESS(status)) {
+    if (IsBigEndian()) {
+      uint16_t* dst = reinterpret_cast<uint16_t*>(*buf);
+      SwapBytes(dst, dst, len);
+    }
+    args.GetReturnValue().Set(
+      Buffer::Copy(env->isolate(),
+                  reinterpret_cast<char*>(*buf),
+                  ts_obj_length << 1).ToLocalChecked());
+    goto cleanup;
+  }
+
+ error:
+  env->ThrowError(u_errorName(status));
+
+ cleanup:
+  if (to_conv != nullptr)
+    ucnv_close(to_conv);
+}
+
+// Convert to a named encoding from UCS2
+// args[0] is the name of the encoding being converted to
+// args[1] must be a buffer instance
+void ConvertFromUcs2(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsString());
+  THROW_AND_RETURN_UNLESS_BUFFER(env, args[1]);
+
+  Utf8Value name(env->isolate(), args[0]);
+  SPREAD_ARG(args[1], ts_obj);
+
+  UConverter* to_conv = nullptr;
+  UErrorCode status = U_ZERO_ERROR;
+
+  UChar* source = nullptr;
+  MaybeStackBuffer<UChar> swapspace;
+  const size_t length = ts_obj_length >> 1;
+
+  if (IsLittleEndian()) {
+    source = reinterpret_cast<UChar*>(ts_obj_data);
+  } else {
+    swapspace.AllocateSufficientStorage(length);
+    source = static_cast<UChar*>(malloc(ts_obj_length >> 1));
```

Comment:
`source` is overridden below with `*swapspace`, I think this line does nothing but leak memory.